### PR TITLE
feature: Add typescript to codebase

### PR DESCRIPTION
**Pull request by jeswr**
This PR upgrades the codebase to use typescript.

Blocked until #90  is merged

### CODE REVIEW

File: src/ComunicaEngine.ts [hunk 1]
```diff
@@ -1,24 +1,69 @@
...
+// Change over to this type once https://github.com/microsoft/TypeScript/issues/47208#issuecomment-1014128087 is resolved
+// export type RawDataSources = DataSource | Promise<RawDataSources> | RawDataSources[];
+export type MaybePromiseArray<T> = T | T[] | Promise<T>;
+export type NestPromiseArray<T> = MaybePromiseArray<MaybePromiseArray<MaybePromiseArray<MaybePromiseArray<MaybePromiseArray<T>>>>>
+export type RawDataSources = NestPromiseArray<DataSource>;
+
+/**
+ * Settings for the ldflex-comunica engine
+ */
+export interface IEngineSettings {
```
Comment (RubenVerborgh) [inline src/ComunicaEngine.ts:18]: Maybe drop the `I`?

File: src/ComunicaEngine.ts [hunk 2]
```diff
@@ -1,24 +1,69 @@
...
+   * @default {}
+   */
+  options?: any;
+}
 
 /**
  * Asynchronous iterator wrapper for the Comunica SPARQL query engine.
  */
 export default class ComunicaEngine {
+  private _sources: Promise<DataSources>;
```
Comment (RubenVerborgh) [inline src/ComunicaEngine.ts:44]: In the future, we can drop the underscore from such private members. But there are still JavaScript consumers for now, so might want to hold off. (I use the `_` in JS to signal private, but TypeScript obsoletes that.)

**Review by RubenVerborgh — APPROVED** [review]
Good stuff, @jeswr, this is great!

File: src/ComunicaEngine.ts (hunk 2 above)
Comment (jeswr) [reply src/ComunicaEngine.ts:44]: Yep - I am one such person (for instance overriding `_engine` before the custom engine option was added). I'll add a comment to change the naming in the release of the next *major* version.

**Pull request merged**